src/trait-wc-gateway-cp-subscriptions.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+require_once(dirname(__FILE__) . '/ApiKey.php');
+require_once dirname(__FILE__) . '/trait-wc-citypay-api.php';
+
+trait WC_Gateway_CP_Subscriptions
+{
+    use WC_CP_API;
+
+    /**
+     *  Checks if subscriptions are enabled on the site.
+     * @return bool
+     */
+    public function is_subscriptions_enabled()
+    {
+        return class_exists('WC_Subscriptions') && $this->cp_subscriptions == 'yes';
+    }
+
+    /**
+     * Initialize subscription support and hooks.
+     * @return void
+     */
+    public function init_subscriptions()
+    {
+        if (!$this->is_subscriptions_enabled()) {
+            return;
+        }
+
+        $this->supports = array_merge(
+            $this->supports,
+            array(
+                'subscriptions',
+                'subscription_cancellation',
+                'subscription_reactivation',
+                'subscription_suspension',
+                'subscription_amount_changes',
+                'subscription_date_changes'
+            )
+        );
+
+        // subscription actions
+        add_action('woocommerce_scheduled_subscription_payment_' . $this->id, array($this, 'scheduled_subscription_payment'), 10, 2);
+        add_action('woocommerce_subscription_failing_payment_method_updated_' . $this->id, [$this, 'update_failing_payment_method'], 10, 2);
+    }
+
+    /**
+     * @param $amount_to_charge // the amount to charge
+     * @param $renewal_order // A WC_Order object created to record the renewal payment.
+     * @return void
+     * @throws Exception
+     */
+    public function scheduled_subscription_payment($amount_to_charge, $renewal_order)
+    {
+        try {
+
+            $this->debugLog('WC_Gateway_CP_Subscriptions::scheduled_subscription_payment()');
+
+            $renewal_order_id = $renewal_order->get_id();
+            $subscriptions =  wcs_get_subscriptions_for_renewal_order($renewal_order_id);
+            $subscription_id = array_values($subscriptions)[0]->get_id();  // we are assuming that we just accept one subscription
+
+            $this->debugLog('scheduled_subscription_payment subscription_id:' . $subscription_id );
+
+            $merchant_id = $this->get_merchant_id();
+            $accountNo = get_post_meta($subscription_id, 'AccountNo', true);
+            $account = $this->account_retrieval($accountNo);
+
+            $token = array_values($account['CardHolderAccount']['cards'])[0]['token'];
+
+            $charge_body = [
+                "amount" => $amount_to_charge,
+                "identifier" => "renewal-" . $merchant_id . $subscription_id . $renewal_order_id,
+                "merchantid" => (int) $merchant_id ,
+                "token" => $token,
```

Comment:
should this not include subscription_id in the charge body?

Reply:
It should. I've just added it.